addons/binding/org.openhab.binding.bigassfan/src/main/java/org/openhab/binding/bigassfan/handler/BigAssFanHandler.java

```diff
@@ -0,0 +1,1061 @@
+        private ScheduledExecutorService scheduledExecutorService = Executors.newScheduledThreadPool(1);
+        private ScheduledFuture<?> listenerJob;
+
+        private final long FAN_LISTENER_DELAY = 2L;
+        private boolean terminate;
+
+        private ConnectionManager conn;
+
+        private Runnable fanListenerRunnable = new Runnable() {
+            @Override
+            public void run() {
+                try {
+                    listener();
+                } catch (Exception e) {
+                    logger.warn("FanListener had unhandled exception: ", e.getMessage());
+                }
+            }
+        };
+
+        public FanListener() {
+            conn = new ConnectionManager();
+        }
+
+        public void startFanListener() {
+            conn.connect();
+            conn.scheduleConnectionMonitorJob();
+
+            if (listenerJob == null) {
+                terminate = false;
+                logger.debug("Starting fan listener in {} sec for {} at {}", FAN_LISTENER_DELAY, thingID(), ipAddress);
+                listenerJob = scheduledExecutorService.schedule(fanListenerRunnable, FAN_LISTENER_DELAY,
+                        TimeUnit.SECONDS);
+            }
+        }
+
+        public void stopFanListener() {
+            if (listenerJob != null) {
+                logger.debug("Stopping fan listener for {} at {}", FAN_LISTENER_DELAY, thingID(), ipAddress);
+                terminate = true;
+                listenerJob.cancel(true);
+                listenerJob = null;
+            }
+            conn.cancelConnectionMonitorJob();
+            conn.disconnect();
+        }
+
+        public void send(String command) {
+            if (!conn.isConnected()) {
+                logger.info("Unable to send message; no connection to fan. Trying to reconnect: {}", command);
+                conn.connect();
+                if (!conn.isConnected()) {
+                    return;
+                }
+            }
+
+            logger.debug("Sending command to fan: {}", command);
+            byte[] buffer;
+            try {
+                buffer = command.getBytes(CHARSET);
+            } catch (UnsupportedEncodingException e) {
+                logger.warn("Unable to convert to string using {} charset: {}", CHARSET, e.getMessage());
+                return;
+            }
+            try {
+                conn.write(buffer);
+            } catch (IOException e) {
+                logger.warn("IO exception writing message to socket: {}", e.getMessage());
+                conn.disconnect();
+            } catch (Exception e) {
+                logger.warn("Unexpected exception writing message to socket: {}", e.getMessage());
+                conn.disconnect();
+            }
+        }
+
+        private void listener() {
+            logger.debug("Fan listener thread is running");
+
+            // XXX Uncomment for testing
+            // testListener();
```

Comment:
Test code can be added in a separate project see for example `org.openhab.binding.feed.test`